src/app/content/utils/canonicalUrl.ts

```diff
@@ -29,7 +29,7 @@ export async function getCanonicalUrlParams(
   let canonicalPageId = pageId;
   let done = false;
   let canonicalBook;
-  let treeSection;
+  let canonicalBookWithPage;
 
   while (canonicalMap.length && !done) {
```

Comment:
`canonicalMap` is an array either of maps that have been specified in canonicalBookMap.ts or an empty default map. it's only undefined if the loaded book doesn't exist in the books config. `canonicalPageId` is the current page id and `canonicalBookWithPage` is undefined.

while there is a map array, loop through each map. set `canonicalBook` to the book id of the map being checked and update `canonicalPageId` to the page specified in the page map (if found). set `treeSection` to the tree node corresponding to the canonical page.

if no tree section was found but a `canonicalBookWithPage` was previously set, end the loop and use that `canonicalBookWithPage` to generate the url (think I was seeing this happen for pages with multiple canonical books specified but where not all those books had a matching page). else if a tree section was found, update `canonicalBookWithPage` to use it.

next check to see if the `canonicalBook` has its own map (happens with precalc 2e for example). if it does not or if it's the same map we just checked, stop the loop. throw an error if we've encountered a loop within the maps themselves. otherwise, set `canonicalMap` to the new map and start over.

once we're past the loop, return a canonical URL using `canonicalBookWithPage` if defined or using the current page's URL. return null if no map was ever defined (see discussion below on this).